### PR TITLE
Spec asynchronous uploads

### PR DESCRIPTION
**Pull request by richvdh**
Spec for https://github.com/matrix-org/matrix-spec-proposals/pull/2246

Preview: https://pr1499--matrix-spec-previews.netlify.app

### CODE REVIEW

File: data/api/client-server/content-repo.yaml [hunk 1]
```diff
@@ -101,7 +100,156 @@ paths:
...
+        - in: header
+          name: Content-Type
+          type: string
+          description: The content type of the file being uploaded
+          x-example: "application/pdf"
+        - in: query
+          type: string
+          x-example: "War and Peace.pdf"
+          name: filename
+          description: The name of the file being uploaded
```
Comment (richvdh) [inline data/api/client-server/content-repo.yaml:121]: TBC, per https://github.com/matrix-org/matrix-spec-proposals/pull/2246/files#r1176474108

File: data/api/client-server/content-repo.yaml [hunk 2]
```diff
@@ -101,7 +101,170 @@ paths:
             "$ref": "definitions/errors/rate_limited.yaml"
       tags:
         - Media
-  "/download/{serverName}/{mediaId}":
+  "/media/v3/upload/{serverName}/{mediaId}":
+    put:
+      summary: Upload content to an `mxc://` URI that was created earlier.
+      description: |-
+        This endpoint permits uploading content to an `mxc://` URI that was created
+        earlier via [POST /_matrix/media/v1/create](/client-server-api/#post_matrixmediav3create).
```
Comment (sumnerevans) [inline data/api/client-server/content-repo.yaml:109]:
```suggestion
        earlier via [POST /_matrix/media/v1/create](/client-server-api/#post_matrixmediav1create).
```

File: data/api/client-server/content-repo.yaml [hunk 3]
```diff
@@ -101,7 +101,170 @@ paths:
...
+          examples:
+            application/json: {}
+        403:
+          description: |-
+            The user does not have permission to upload the content. Some reasons for this error include:
+
+            - The server does not permit the file type.
+            - The user has reached a quota for uploaded content.
+            - The request comes from a different user than the one that called
+              [POST /_matrix/media/v1/create](/client-server-api/#post_matrixmediav3create).
```
Comment (sumnerevans) [inline data/api/client-server/content-repo.yaml:160]:
```suggestion
              [POST /_matrix/media/v1/create](/client-server-api/#post_matrixmediav1create).
```

File: data/api/client-server/content-repo.yaml [hunk 4]
```diff
@@ -156,10 +337,17 @@ paths:
...
+            application/json: {
+              "errcode": "M_NOT_YET_UPLOADED",
+              "error": "Content has not yet been uploaded"
+            }
           schema:
-            "$ref": "definitions/errors/rate_limited.yaml"
+            "$ref": "definitions/errors/error.yaml"
       tags:
         - Media
   "/download/{serverName}/{mediaId}/{fileName}":
```
Comment (uhoreg) [inline data/api/client-server/content-repo.yaml:353]:
```suggestion
  "/media/v3/download/{serverName}/{mediaId}/{fileName}":
```

**Review by uhoreg — APPROVED** [review]
lgtm other than a nit

**Pull request merged**